### PR TITLE
#969: Use new CTPG parser in java vm

### PR DESCRIPTION
**Pull request by tomuben**
related to https://github.com/exasol/script-languages-release/issues/969

**Please note:** The unit tests at https://github.com/exasol/script-languages/blob/master/exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc are being re-used to test identical behavior of both parser: The old (named legacy) parser, and the new (based on CTPG) parser.

### CODE REVIEW

File: exaudfclient/exaudfclient.cc [hunk 1]
```diff
@@ -163,7 +163,7 @@ int main(int argc, char **argv) {
     } else if (strcmp(argv[2], "lang=java")==0)
     {
 #ifdef ENABLE_JAVA_VM
-        vmMaker = [&](){return new SWIGVMContainers::JavaVMach(false, swigFactory);};
+        vmMaker = [&](){return new SWIGVMContainers::JavaVMach(false, swigFactory, false);};
```
Comment (tomuben) [inline exaudfclient/exaudfclient.cc:166]: Proper forwarding of the parameter will be handled in another PR

File: exaudfclient/base/javacontainer/test/cpp/javacontainer_ctpg_test.cc [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            std::make_pair("\\;-Dhttp.agent=ABCDEF", ";-Dhttp.agent=ABCDEF"),
+            std::make_pair("-Dhttp.agent=ABCDEF\\n", "-Dhttp.agent=ABCDEF\n"),
+            std::make_pair("-Dhttp.agent=ABCDEF\\r", "-Dhttp.agent=ABCDEF\r"),
+            std::make_pair("-Dhttp.agent=ABCDEF\\;", "-Dhttp.agent=ABCDEF;"),
+            std::make_pair("\\ -Dhttp.agent=ABCDEF", "-Dhttp.agent=ABCDEF"),
+            std::make_pair("\\t-Dhttp.agent=ABCDEF", "-Dhttp.agent=ABCDEF"),
+            std::make_pair("\\f-Dhttp.agent=ABCDEF", "-Dhttp.agent=ABCDEF"),
+            std::make_pair("\\v-Dhttp.agent=ABCDEF", "-Dhttp.agent=ABCDEF")
+        };
+
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/test/cpp/javacontainer_ctpg_test.cc:57]: Do we have a test for escaped whitespace in the middle of a option, because this was one of the cases that was as bug reported

-Dhttp.agent=ABC\ DEF;

File: exaudfclient/base/javacontainer/javacontainer_impl.cc [hunk 3]
```diff
@@ -64,6 +58,21 @@ JavaVMImpl::JavaVMImpl(bool checkOnly, bool noJNI, SwigFactory& swigFactory)
     }
 }
 
+void JavaVMImpl::parseScriptOptions(JavaScriptOptions::ScriptOptionsParser & scriptOptionsParser, SwigFactory& swigFactory) {
+    JavaScriptOptions::Extractor extractor(scriptOptionsParser, swigFactory);
+
+    DBG_FUNC_CALL(cerr,extractor.extract(m_scriptCode));  // To be called before scripts are imported. Otherwise, the script classname from an imported script could be used
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/javacontainer_impl.cc:64]: Does the comment still makes sense, I don't see the import here, doesn't the extract handle this, too

File: exaudfclient/base/javacontainer/script_options/parser_ctpg.cc [hunk 4]
```diff
@@ -1,9 +1,147 @@
...
+    for (const auto option_loc: option_locations) {
+        m_scriptCode.erase(option_loc.pos, option_loc.len);
+    }
+    return std::move(m_scriptCode);
+}
+
+void ScriptOptionLinesParserCTPG::parse() {
+    if (m_needParsing) {
+        if(!m_foundOptions.empty()) {
+            throw std::logic_error("F-UDF-CL-SL-JAVA-1620 Internal error. Parser result is not empty.");
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/script_options/parser_ctpg.cc:99]: That should not happen, except we screwed up, I guess, maybe we should mention that this indicates a bug and user should open a ticket

File: exaudfclient/base/javacontainer/script_options/parser_ctpg.cc [hunk 5]
```diff
@@ -1,9 +1,147 @@
...
+
+        //Check for unknown options
+        for (const auto & option: m_foundOptions) {
+            if (m_keywords.jarKeyword() != option.first &&
+                m_keywords.scriptClassKeyword() != option.first  &&
+                m_keywords.importKeyword() != option.first  &&
+                m_keywords.jvmKeyword() != option.first) {
+                    std::stringstream ss;
+                    ss << "F-UDF-CL-SL-JAVA-1622 " << "Unexpected option: " << option.first;
+                    throw std::invalid_argument(ss.str());
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/script_options/parser_ctpg.cc:117]: Hm, is this really a exception, with old one the option simply would be ignored and stays in the text

File: exaudfclient/base/javacontainer/script_options/parser_ctpg.cc [hunk 6]
```diff
@@ -1,9 +1,147 @@
...
+    }
+}
+
+void ScriptOptionLinesParserCTPG::parseForSingleOption(const std::string key, std::function<void(const std::string &option)> callback) {
+    parse();
+    const auto optionIt = m_foundOptions.find(key);
+    if (optionIt != m_foundOptions.end()) {
+        if (optionIt->second.size() != 1) {
+            std::stringstream ss;
+            ss << "F-UDF-CL-SL-JAVA-1628 found " << optionIt->second.size() <<  key << " options" << std::endl;
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/script_options/parser_ctpg.cc:129]: I think there is space between size and key missing

File: exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+, m_swigFactory(swigFactory)
+, m_metaData()
+, m_keywords(keywords) {}
+
+void ScriptImporter::importScript(std::string & scriptCode, ctpg_parser::options_map_t & options) {
+    const auto optionIt = options.find(std::string(m_keywords.importKeyword()));
+    if (optionIt != options.end()) {
+        m_importedScriptChecksums.addScript(scriptCode.c_str());
+        //Sort options from first in script to last in script
+        std::sort(optionIt->second.begin(), optionIt->second.end(),
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc:30]: This is inplace sorting or

Comment (tkilias) [reply exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc:30]: Thats why also a non const ref for options

File: exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        replacedScripts.reserve(optionIt->second.size());
+        //In order to continue compatibility with legacy implementation we must collect import scripts in forward direction
+        //but then replace in reverse direction (in order to keep consistency of positions)
+        for (const auto & option: optionIt->second) {
+            const char *importScriptCode = findImportScript(option.value);
+            std::string importScriptCodeStr;
+            if (m_importedScriptChecksums.addScript(importScriptCode)) {
+                // Script has not been imported yet
+                // If this imported script contains %import statements
+                // they will be resolved in the next recursion.
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc:51]: This time it is really recursion, could we get recursion depth issues?

File: exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+                } catch(const ExecutionGraph::OptionParserException & ex) {
+                    Utils::rethrow(ex, "F-UDF-CL-SL-JAVA-1630");
+                }
+                importScriptCodeStr.assign(importScriptCode);
+                importScript(importScriptCodeStr, newOptions);
+            }
+            ReplacedScripts replacedScript = {.script = std::move(importScriptCodeStr), .origPos = option.idx_in_source, .origLen = option.size };
+            replacedScripts.push_back(std::move(replacedScript));
+        }
+        for (auto optionIt = replacedScripts.rbegin(); optionIt != replacedScripts.rend(); optionIt++) {
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc:64]: Maybe add comment, iterate in reverse because we need to replace from the end

File: exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+                    ExecutionGraph::OptionsLineParser::CTPG::parseOptions(importScriptCode, newOptions);
+                } catch(const ExecutionGraph::OptionParserException & ex) {
+                    Utils::rethrow(ex, "F-UDF-CL-SL-JAVA-1630");
+                }
+                importScriptCodeStr.assign(importScriptCode);
+                importScript(importScriptCodeStr, newOptions);
+            }
+            ReplacedScripts replacedScript = {.script = std::move(importScriptCodeStr), .origPos = option.idx_in_source, .origLen = option.size };
+            replacedScripts.push_back(std::move(replacedScript));
+        }
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc:63]: What is with the else branch, script already imported, we need to remove the option from the code

File: exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+                // they will be resolved in the next recursion.
+                ctpg_parser::options_map_t newOptions;
+                try {
+                    ExecutionGraph::OptionsLineParser::CTPG::parseOptions(importScriptCode, newOptions);
+                } catch(const ExecutionGraph::OptionParserException & ex) {
+                    Utils::rethrow(ex, "F-UDF-CL-SL-JAVA-1630");
+                }
+                importScriptCodeStr.assign(importScriptCode);
+                importScript(importScriptCodeStr, newOptions);
+            }
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc:60]: What is with the else branch, script already imported, we need to remove the option from the code.

Comment (tkilias) [reply exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc:60]: Oh this handle by the init value of importScriptCodeStr, puh

File: exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc [hunk 12]
```diff
@@ -369,7 +369,10 @@ TEST(JavaContainer, import_script_script_class_option_ignored) {
     EXPECT_EQ(vm.getJavaVMInternalStatus().m_localClasspath, "/tmp");
     const std::string expected_script_code =
         "package com.exasol;\r\n"
-        "%scriptclass com.exasol.udf_profiling.UdfProfiler;\n"
+#ifndef USE_CTPG_PARSER //The parsers behave differently: The legacy parser incorrectly keeps imported scriptclass options
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc:372]: Ohhhh, another bug

File: exaudfclient/base/javacontainer/test/cpp/javacontainer_ctpg_test.cc (hunk 2 above)
Comment (tomuben) [reply exaudfclient/base/javacontainer/test/cpp/javacontainer_ctpg_test.cc:57]: This bug is not fixed yet. 
My plan was to fix https://github.com/exasol/script-languages-release/issues/878 in another PR, in order to keep a clean git history.

In the scope of #878 I will extend the test here.

File: exaudfclient/base/javacontainer/javacontainer_impl.cc (hunk 3 above)
Comment (tomuben) [reply exaudfclient/base/javacontainer/javacontainer_impl.cc:64]: True. I moved the comment to the correct location

File: exaudfclient/base/javacontainer/script_options/parser_ctpg.cc (hunk 4 above)
Comment (tomuben) [reply exaudfclient/base/javacontainer/script_options/parser_ctpg.cc:99]: Done

File: exaudfclient/base/javacontainer/script_options/parser_ctpg.cc (hunk 5 above)
Comment (tomuben) [reply exaudfclient/base/javacontainer/script_options/parser_ctpg.cc:117]: I could have implemented the same behavior as in the old legacy parser, but then I would need to filter the known options in ScriptOptionLinesParserCTPG::getScriptCode() (where we remove the options from script code).

My thought was to keep the implementation simple, and also throwing an explicit error here is better than getting a strange syntax error back from the Java compiler later.

File: exaudfclient/base/javacontainer/script_options/parser_ctpg.cc (hunk 6 above)
Comment (tomuben) [reply exaudfclient/base/javacontainer/script_options/parser_ctpg.cc:129]: Done

File: exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc (hunk 7 above)
Comment (tomuben) [reply exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc:30]: Exactly

File: exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc (hunk 8 above)
Comment (tomuben) [reply exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc:51]: We could add an counter and limit the number of nested scripts to  maybe 100. What do you think?

File: exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc (hunk 9 above)
Comment (tomuben) [reply exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc:64]: There is a comment for that logic at line 44

File: exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc (hunk 12 above)
Comment (tomuben) [reply exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc:372]: Yes

Comment (tomuben) [reply exaudfclient/base/javacontainer/test/cpp/javacontainer_test.cc:372]: I discovered this when running the unit tests on the new parser :)

File: exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc (hunk 11 above)
Comment (tomuben) [reply exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc:60]: Yes,I believe, besides the parser itself,  this function is one of the most difficult ones of the new parser implementation

File: exaudfclient/base/javacontainer/script_options/parser_ctpg.cc (hunk 5 above)
Comment (tkilias) [reply exaudfclient/base/javacontainer/script_options/parser_ctpg.cc:119]: Oh, you are right we know the language and that every option line that we encounter will be wrong.

File: exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc (hunk 9 above)
Comment (tkilias) [reply exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc:75]: But you did a whole bunch between the comment and reverse iteration

File: exaudfclient/base/javacontainer/script_options/test/ctpg_script_importer_test.cc [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    ss << "%import " << sc_scriptName << currentIdx << ";something";
+    static std::string ret;
+    ret = ss.str();
+    return ret.c_str();
+}
+
+TEST(ScriptImporterTest, max_recursion_depth) {
+
+    size_t currentIdx = 0;
+    SwigFactoryTestImpl swigFactoryTest([&](const char* scriptKey) {
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/script_options/test/ctpg_script_importer_test.cc:34]: Nasty, you generate the Code when. It gets requested. Took a moment to see where this runs in a loop. Maybe a comment can help

File: exaudfclient/base/javacontainer/script_options/test/swig_factory_test.h [hunk 14]
```diff
@@ -0,0 +1,20 @@
+#ifndef SWIG_FACTORY_TEST_H
+#define SWIG_FACTORY_TEST_H 1
+
+#include <string>
+#include <map>
+#include "base/swig_factory/swig_factory.h"
+#include <functional>
+
+struct SwigFactoryTestImpl : public SWIGVMContainers::SwigFactory {
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/script_options/test/swig_factory_test.h:9]: And, a thought I had the factory we give the parser should generate a object that is only able return module content this way we can reuse it for the potential new client

File: exaudfclient/base/javacontainer/test/cpp/javavm_test.cc [hunk 15]
```diff
@@ -32,6 +32,6 @@ void JavaVMTest::run(std::string scriptCode, SwigFactoryTestImpl & swigFactory)
     javaVMInternalStatus.m_classpath = javaVMImpl.m_classpath;
     javaVMInternalStatus.m_jvmOptions = javaVMImpl.m_jvmOptions;
     javaVMInternalStatus.m_needsCompilation = javaVMImpl.m_needsCompilation;
-    delete script_code;
+    ::free(script_code);
 }
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/test/cpp/javavm_test.cc:36]: Huh?

Comment (tomuben) [reply exaudfclient/base/javacontainer/test/cpp/javavm_test.cc:41]: Reported by asan :)
`script_code` was allocated with `::malloc` (strdup) but deallocated with `delete`.
If you allocate with `::malloc` you should deallocated with `::free`.
If you allocate with `new` you should delete with `delete`

File: exaudfclient/base/javacontainer/script_options/test/ctpg_script_importer_test.cc (hunk 13 above)
Comment (tomuben) [reply exaudfclient/base/javacontainer/script_options/test/ctpg_script_importer_test.cc:34]: Added a comment

File: exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc (hunk 8 above)
Comment (tomuben) [reply exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc:62]: Added a check for max number of recursions and throwing an exception if number exceeds limit

File: .github/workflows/check_bazel_tests.yml [hunk 16]
```diff
@@ -26,7 +26,7 @@ jobs:
           sudo apt-get install -y openjdk-11-jdk libzmq3-dev
       - name: Build
         run: |
-          bazel build --lockfile_mode=off --config no-tty -c dbg --config python --config java --config fast-binary --verbose_failures
+          bazel build --lockfile_mode=off --config no-tty -c dbg --config python --config fast-binary --verbose_failures
```
Comment (tkilias) [inline .github/workflows/check_bazel_tests.yml:29]: we should also test the build of java

File: exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+                } catch(const ExecutionGraph::OptionParserException & ex) {
+                    Utils::rethrow(ex, "F-UDF-CL-SL-JAVA-1630");
+                }
+                importScriptCodeStr.assign(importScriptCode);
+                importScript(importScriptCodeStr, newOptions, recursionDepth + 1);
+            }
+            ReplacedScripts replacedScript = {.script = std::move(importScriptCodeStr), .origPos = option.idx_in_source, .origLen = option.size };
+            replacedScripts.push_back(std::move(replacedScript));
+        }
+        //No replace the imported script bodies from end to start. Doing it in forward order would invalidate the offsets of later import scripts.
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/script_options/parser_ctpg_script_importer.cc:75]:
```suggestion
        //Now replace the imported script bodies from end to start. Doing it in forward order would invalidate the offsets of later import scripts.
```

File: .github/workflows/check_bazel_tests.yml (hunk 16 above)
Comment (tomuben) [reply .github/workflows/check_bazel_tests.yml:29]: I remember that I tried that once and then I gave up because of the dependencies....

Comment (tomuben) [reply .github/workflows/check_bazel_tests.yml]: Readded --config java

**Review by tkilias — APPROVED** [review]
